### ISSUE TITLE
countNonZero returns wrong counts for unit8_t arrays with large dimensions (IPP)

### ISSUE THREAD

**Issue by D-Alex** [opencv/opencv#17453]
- OpenCV => 4.3
- Operating System / Platform =>Mac OS Catalina + OpenCL + IPP
- Compiler => clang

##### Detailed description

countNonZero is returning wrong counts for large unit8_t arrays. 

The issue seems to be related to ippiCountInRange_8u_C1R which is called behind the scene returning a wrong inverse count.

##### Steps to reproduce

```.cpp
std::cout << cv::countNonZero(cv::Mat::zeros(1,524190-1,CV_8UC1));     // returns 0
std::cout << cv::countNonZero(cv::Mat::zeros(1,524190,CV_8UC1));        // returns 65536
std::cout << cv::countNonZero(cv::Mat::zeros(1,524190+1,CV_8UC1));    // returns 65536
```

**Comment by mshabunin** [comment]
Looks like this problem can be reproduced only on OSX and Win32. Perhaps we can disable IPP path for these platforms temporarily.

**Comment by diablodale** [comment]
Hi all. I can reproduce this exact fail using the current 4.5.x master branch e726ff3296b9e382371a778ad807d37a2e49b979 with VS2019 and a **64-bit** compile target.
I think the #ifdef added in the PR is too narrow to fully resolve this issue.
Test case `Core/CountNonZeroBig./1, where GetParam() = (0, 524190x1)` fails.

#### Cause

https://github.com/opencv/opencv/blob/e726ff3296b9e382371a778ad807d37a2e49b979/modules/core/src/count_non_zero.dispatch.cpp#L65

That code is looking for Apple *or* (Microsoft MSVC compiler with 32-bit compile target). `_M_IX86` is *not* defined for Windows 64-bit compile target. https://docs.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=msvc-160

The original report above is broader. It is all of Windows. Or perhaps all of MSVC compiling for Windows. I don't see anything in the bug report distinguishing that it is only Windows 32-bit compile targets. @mshabunin, you investigated the affected groups. Is this true?

#### Recommendation

Therefore, I recommend the #ifdef be broader to cover all Windows x86 and x64 targets. It is unsafe to use only `_WIN32` because the MSVC compiler also targets Windows ARM compile targets. Therefore, I recommend `_M_X64` or `_M_AMD64` those two are equivalent https://docs.microsoft.com/en-us/archive/blogs/reiley/macro-revisited

Leading to a suggested ifdef of...

```c++
#if defined __APPLE__ || (defined _MSC_VER && (defined _M_IX86 || defined _M_X64))
```

Using that suggested change, the master branch compile and the all test cases pass `[  PASSED  ] 11609 tests`

If everyone agrees, I can make another PR.

**Comment by alalek** [comment]
[Can't reproduce](http://pullrequest.opencv.org/buildbot/builders/master-win64-vc16/builds/555) (nightly build, Dec 1 2020, CPU is i7-6700K).
Please collect extra information from test run (`opencv_test_core`):
```
...
CPU features: SSE SSE2 SSE3 *SSE4.1 *SSE4.2 *FP16 *AVX *AVX2 *AVX512-SKX?
Intel(R) IPP version: ippIP AVX2 (l9) 2020.0.0 Gold (-) Oct 21 2019
...
[----------] 6 tests from Core/CountNonZeroBig
[ RUN      ] Core/CountNonZeroBig./0, where GetParam() = (0, 1x524190)
[       OK ] Core/CountNonZeroBig./0 (4 ms)
[ RUN      ] Core/CountNonZeroBig./1, where GetParam() = (0, 524190x1)
[       OK ] Core/CountNonZeroBig./1 (0 ms)
[ RUN      ] Core/CountNonZeroBig./2, where GetParam() = (0, 3840x2160)
[       OK ] Core/CountNonZeroBig./2 (5 ms)
[ RUN      ] Core/CountNonZeroBig./3, where GetParam() = (5, 1x524190)
[       OK ] Core/CountNonZeroBig./3 (3 ms)
[ RUN      ] Core/CountNonZeroBig./4, where GetParam() = (5, 524190x1)
[       OK ] Core/CountNonZeroBig./4 (1 ms)
[ RUN      ] Core/CountNonZeroBig./5, where GetParam() = (5, 3840x2160)
[       OK ] Core/CountNonZeroBig./5 (26 ms)
[----------] 6 tests from Core/CountNonZeroBig (39 ms total)
```

**Comment by diablodale** [comment]
info below

#### Setup

* master e726ff3296b9e382371a778ad807d37a2e49b979
* MSVS2019 Community v16.8.2
* windows amd64 (x64) compile target
* Debug build

#### Repro

1. Build opencv
2. Run `opencv_test_cored.exe --gtest_filter=*CountNonZeroBig*`

#### Result

```
CTEST_FULL_OUTPUT
OpenCV version: 4.5.1-pre
OpenCV VCS version: 4.5.0-254-ge726ff3296-dirty
Build type: Debug
Compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.28.29333/bin/Hostx64/x64/cl.exe  (ver 19.28.29334.0)
Parallel framework: ms-concurrency (nthreads=8)
CPU features: SSE SSE2 SSE3 SSSE3 SSE4.1 POPCNT SSE4.2 AVX *FP16 *AVX2? *AVX512-SKX?
Intel(R) IPP version: ippIP SSE4.2 (y8) 2020.0.0 Gold (-) Oct 21 2019
[ INFO:0] global ..\modules\core\src\ocl.cpp (1176) cv::ocl::haveOpenCL Initialize OpenCL runtime...
[ INFO:0] global ..\modules\core\src\ocl.cpp (1182) cv::ocl::haveOpenCL OpenCL: found 2 platforms
[ INFO:0] global ..\modules\core\src\ocl.cpp (974) cv::ocl::OpenCLExecutionContext::Impl::getInitializedExecutionContext OpenCL: initializing thread execution context
[ INFO:0] global ..\modules\core\src\ocl.cpp (984) cv::ocl::OpenCLExecutionContext::Impl::getInitializedExecutionContext OpenCL: creating new execution context...
[ INFO:0] global ..\modules\core\src\ocl.cpp (1012) cv::ocl::OpenCLExecutionContext::Impl::getInitializedExecutionContext OpenCL: context is not available/disabled
OpenCL is disabled
TEST: Skip tests with tags: 'mem_6gb', 'verylong', 'debug_verylong'
Note: Google Test filter = *CountNonZeroBig*
[==========] Running 6 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 6 tests from Core/CountNonZeroBig
[ RUN      ] Core/CountNonZeroBig./0, where GetParam() = (0, 1x524190)
[       OK ] Core/CountNonZeroBig./0 (4 ms)
[ RUN      ] Core/CountNonZeroBig./1, where GetParam() = (0, 524190x1)
..\modules\core\test\test_countnonzero.cpp(287): error: Expected equality of these values:
  0
  cv::countNonZero(cv::Mat::zeros(sz, type))
    Which is: 65536
[  FAILED  ] Core/CountNonZeroBig./1, where GetParam() = (0, 524190x1) (2 ms)
[ RUN      ] Core/CountNonZeroBig./2, where GetParam() = (0, 3840x2160)
[       OK ] Core/CountNonZeroBig./2 (14 ms)
[ RUN      ] Core/CountNonZeroBig./3, where GetParam() = (5, 1x524190) 
[       OK ] Core/CountNonZeroBig./3 (7 ms)
[ RUN      ] Core/CountNonZeroBig./4, where GetParam() = (5, 524190x1) 
[       OK ] Core/CountNonZeroBig./4 (4 ms)
[ RUN      ] Core/CountNonZeroBig./5, where GetParam() = (5, 3840x2160)
[       OK ] Core/CountNonZeroBig./5 (64 ms)
[----------] 6 tests from Core/CountNonZeroBig (128 ms total)

[----------] Global test environment tear-down
[==========] 6 tests from 1 test case ran. (136 ms total)    
[  PASSED  ] 5 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] Core/CountNonZeroBig./1, where GetParam() = (0, 524190x1)

 1 FAILED TEST
```

**Comment by diablodale** [comment]
Two differences I notice in our logs....
* I build `debug` and you `release`
* I build IPP `ippIP SSE4.2` and you `ippIP AVX2`. The build system likely chose mine to be only `SSE4.2` because my build machine doesn't have a cpu with AVX2.

**Comment by alalek** [comment]
@diablodale Thank you!

You are right, both Win32 and Mac platforms run SSE4.2 IPP optimizations only (no AVX2).
Problem is reproduced on Linux with `OPENCV_IPP=sse42`.
```
[ RUN      ] Core/CountNonZeroBig./1, where GetParam() = (0, 524190x1)
/home/alalek/projects/opencv/dev/modules/core/test/test_countnonzero.cpp:287: Failure
Expected equality of these values:
  0
  cv::countNonZero(cv::Mat::zeros(sz, type))
    Which is: 65536
[  FAILED  ] Core/CountNonZeroBig./1, where GetParam() = (0, 524190x1) (1 ms)
```


Prepared fix here: #18986
